### PR TITLE
Linked Time: Make `timeSelection` optional in fob controllers

### PR DESCRIPTION
**Pull request by rileyajones**
* Motivation for features / changes
We would like to be able to enable step selection using a prospective fob. The prospective fob is currently implemented withing the fob_controller_component which has a required `timeSelection` parameter.

* Screenshots of UI changes
None

* Detailed steps to verify changes work correctly (as executed by you)
1) Start tensorboard
2) Go to [http://localhost:6006?enableDataTable=true&enableProspectiveFob=true](http://localhost:6006?enableDataTable=true&enableProspectiveFob=true)
3) View a scalar card
4) Assert that no fobs are shown
5) Enable step selection
6) Assert that a single fob is shown

### CODE REVIEW

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts [hunk 1]
```diff
@@ -2376,6 +2376,24 @@ describe('scalar card', () => {
...
+        const scalarCardFobController = fixture.debugElement.query(
+          By.directive(ScalarCardFobController)
+        ).componentInstance;
+        delete scalarCardFobController.timeSelection;
+        fixture.detectChanges();
+        const fobController = fixture.debugElement.query(
+          By.directive(CardFobComponent)
+        ).componentInstance;
+        expect(fobController.startFobWrapper).not.toBeDefined();
+        expect(fobController.endFobWrapper).not.toBeDefined();
```
Comment (japie1235813) [inline tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts:2395]: Why not use "toBeUnDefined" here?

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts [hunk 2]
```diff
@@ -2376,6 +2376,24 @@ describe('scalar card', () => {
...
+        });
+        const fixture = createComponent('card1');
+        const scalarCardFobController = fixture.debugElement.query(
+          By.directive(ScalarCardFobController)
+        ).componentInstance;
+        delete scalarCardFobController.timeSelection;
+        fixture.detectChanges();
+        const fobController = fixture.debugElement.query(
+          By.directive(CardFobComponent)
+        ).componentInstance;
```
Comment (japie1235813) [inline tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts:2393]: nit: could you have new lines between 2393-2394? Have tests separated into blocks (set up, action, expect) helps with the reading.
I honestly am not sure where is the line between setup and action. line 2390-2391?

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts [hunk 3]
```diff
@@ -2376,6 +2376,24 @@ describe('scalar card', () => {
...
+      it('does not render fobs when no timeSelection is provided', fakeAsync(() => {
+        store.overrideSelector(getMetricsLinkedTimeSelection, {
+          start: {step: 20},
+          end: null,
+        });
+        const fixture = createComponent('card1');
+        const scalarCardFobController = fixture.debugElement.query(
+          By.directive(ScalarCardFobController)
+        ).componentInstance;
+        delete scalarCardFobController.timeSelection;
```
Comment (japie1235813) [inline tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts:2389]: Why do we need to set the time selection and delete it here?

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_fob_controller.ts [hunk 4]
```diff
@@ -74,18 +74,18 @@ export class ScalarCardFobController {
     return this.scale.forward(
       this.minMaxHorizontalViewExtend,
       [0, this.axisSize],
-      this.timeSelection.start.step
+      this.timeSelection?.start.step ?? this.minMaxStep.minStep
```
Comment (japie1235813) [inline tensorboard/webapp/metrics/views/card_renderer/scalar_card_fob_controller.ts:77]: why don't we return empty string if timeSelection is undefined?

File: tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts [hunk 5]
```diff
@@ -266,6 +269,9 @@ export class CardFobControllerComponent {
   }
 
   getDraggingFobStep(): number {
+    if (!this.timeSelection) {
+      throw new Error('TimeSelection is not defined');
```
Comment (japie1235813) [inline tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts:273]: I have the impression that if there's any kind of error in Angular app, all functionality just broken. Can we do something soft here like return null or even handle this in isDraggingLower/Higher?
Actually, not in this PR, I already saw some conflict here.

File: tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts [hunk 6]
```diff
@@ -325,6 +334,9 @@ export class CardFobControllerComponent {
    * fob toggles the feature entirely.
    */
   onFobRemoved(fob: Fob) {
+    if (!this.timeSelection) {
+      return;
+    }
```
Comment (japie1235813) [inline tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts:339]: nit: If there is no timeSelection, there is no fob, and this method should not be triggered. Can we do `this.timeSelection!` in this method or you prefer a check and return?
Same as `stepTyped` method.

**Review by japie1235813 — COMMENTED** [review]
Generally lgtm. Thanks for putting up things together so fast!
Still, I some small comments and will reserve lgtm later.

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts (hunk 2 above)
Comment (rileyajones) [reply tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts:2393]: Sure

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_fob_controller.ts (hunk 4 above)
Comment (rileyajones) [reply tensorboard/webapp/metrics/views/card_renderer/scalar_card_fob_controller.ts:77]: Empty string is not a valid argument for that function.

Screenshot from 2022-10-31 09-46-45

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts (hunk 3 above)
Comment (rileyajones) [reply tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts:2389]: Without #6006 the card fob controller won't exist when time selection is not defined.

File: tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts (hunk 5 above)
Comment (rileyajones) [reply tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts]: Yeah that's true, I was hoping adding a loud failure here would make it more detectable when the app is being worked on. It really shouldn't be possible to drag the fob  when the `timeSelection` is `undefined` so this isn't something I would expect to occur during normal use. Would you prefer I just use a non null assertion in this function?

File: tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts (hunk 6 above)
Comment (rileyajones) [reply tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts]: I really don't like adding non null assertions but it does make sense in this case.

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts (hunk 3 above)
Comment (japie1235813) [reply tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts:2389]: I see. Could you mark a TODO here to update the test in #6006? I feel it weird to use delete after #6006

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts (hunk 2 above)
Comment (japie1235813) [reply tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts:2393]: Would a line between line 2390-2391 make sense?

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_fob_controller.ts (hunk 4 above)
Comment (japie1235813) [reply tensorboard/webapp/metrics/views/card_renderer/scalar_card_fob_controller.ts:77]: oh I mean return empty string for `getAxisPositionFromStartStep` then here will be just `this.timeSelection!.start.step`?

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_fob_controller.ts [hunk 7]
```diff
@@ -74,18 +74,18 @@ export class ScalarCardFobController {
...
     );
   }
 
   getAxisPositionFromEndStep() {
-    if (this.timeSelection.end === null) {
+    if (!this.timeSelection?.end) {
       return null;
```
Comment (japie1235813) [inline tensorboard/webapp/metrics/views/card_renderer/scalar_card_fob_controller.ts:83]: why return null not empty string?

File: tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts (hunk 5 above)
Comment (japie1235813) [reply tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts]: I would prefer throw error when there is significant mistake (data loading) that makes the web app broken. Dragging the fob does not feel like one.
I would prefer have assertion then.

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_fob_controller.ts (hunk 7 above)
Comment (rileyajones) [reply tensorboard/webapp/metrics/views/card_renderer/scalar_card_fob_controller.ts:86]: I have no idea, you wrote this line of code.

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts (hunk 3 above)
Comment (rileyajones) [reply tensorboard/webapp/metrics/views/card_renderer/scalar_card_test.ts:2390]: Sure, I'll add a TODO and refactor the test alongside #6006

File: tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts (hunk 5 above)
Comment (rileyajones) [reply tensorboard/webapp/widgets/card_fob/card_fob_controller_component.ts]: I'll update this to use a non null assertion. If that situation occurred, a null pointer exception would be thrown so the web app would be broken

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_fob_controller.ts (hunk 7 above)
Comment (japie1235813) [reply tensorboard/webapp/metrics/views/card_renderer/scalar_card_fob_controller.ts:86]: oops.

**Review by japie1235813 — APPROVED** [review]

**Pull request merged**